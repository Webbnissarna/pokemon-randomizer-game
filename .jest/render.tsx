import { render as rtlRender } from "@testing-library/react";
import { InitializeColorMode, ThemeProvider } from "theme-ui";

import theme from "../utils/theme";

import "./style.css";

export function render(ui: React.ReactElement) {
  return rtlRender(
    <>
      <InitializeColorMode />
      <ThemeProvider theme={theme}>{ui}</ThemeProvider>
    </>
  );
}
