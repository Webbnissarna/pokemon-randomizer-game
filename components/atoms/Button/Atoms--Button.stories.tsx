import React from "react";
import Button from ".";
import type { ClickEvent, Size, Variant } from "./Button.types";

interface ControlProps {
  size: Size;
  variant: Variant;
  text: string;
  disabled: boolean;
}

export function Default({ size, variant, text, disabled }: ControlProps) {
  const onClick = (event: ClickEvent) => {
    console.log("click", event);
  };

  const sizes: [string, Size][] = [
    ["Small", "sm"],
    ["Medium", "md"],
    ["Large", "lg"],
  ];

  return (
    <div style={{ display: "flex", flexDirection: "column", gap: 20 }}>
      <h1>Custom</h1>
      <Button
        onClick={onClick}
        size={size}
        variant={variant}
        text={text}
        disabled={disabled}
      />

      {sizes.map(([title, buttonSize]) => (
        <React.Fragment key={title}>
          <h1>{title}</h1>
          <Button
            onClick={onClick}
            size={buttonSize}
            variant="positive"
            text="Positive"
          />
          <Button
            onClick={onClick}
            size={buttonSize}
            variant="negative"
            text="Negative"
          />
          <Button
            onClick={onClick}
            size={buttonSize}
            variant="neutral"
            text="Neutral"
          />
          <Button
            onClick={onClick}
            size={buttonSize}
            variant="secondary"
            text="Secondary"
          />
          <Button
            onClick={onClick}
            size={buttonSize}
            variant="tertiary"
            text="Tertiary"
          />

          <Button
            onClick={onClick}
            size={buttonSize}
            variant="positive"
            text="Positive (disabled)"
            disabled={true}
          />
          <Button
            onClick={onClick}
            size={buttonSize}
            variant="negative"
            text="Negative (disabled)"
            disabled={true}
          />
          <Button
            onClick={onClick}
            size={buttonSize}
            variant="neutral"
            text="Neutral (disabled)"
            disabled={true}
          />
          <Button
            onClick={onClick}
            size={buttonSize}
            variant="secondary"
            text="Secondary (disabled)"
            disabled={true}
          />
          <Button
            onClick={onClick}
            size={buttonSize}
            variant="tertiary"
            text="Tertiary (disabled)"
            disabled={true}
          />
        </React.Fragment>
      ))}
    </div>
  );
}

Default.args = {
  size: "sm",
  variant: "positive",
  text: "Hello World",
  disabled: false,
} as ControlProps;

Default.argTypes = {
  size: {
    options: ["sm", "md", "lg"],
    control: { type: "select" },
  },
  variant: {
    options: ["positive", "negative", "neutral", "secondary", "tertiary"],
    control: { type: "radio" },
    defaultValue: "positive",
  },
};
