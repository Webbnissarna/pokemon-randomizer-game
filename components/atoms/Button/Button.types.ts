import type { MouseEvent as ReactMouseEvent } from "react";

export type Variant =
  | "positive"
  | "negative"
  | "neutral"
  | "secondary"
  | "tertiary";

export type Size = "sm" | "md" | "lg";

export type ClickEvent = ReactMouseEvent<HTMLButtonElement, MouseEvent>;

export interface ButtonProps {
  text: string;
  variant: Variant;
  size: Size;
  onClick: (event: ClickEvent) => void;
  disabled?: boolean;
}
