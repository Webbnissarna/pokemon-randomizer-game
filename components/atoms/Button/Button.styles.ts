import { ThemeUIStyleObject } from "theme-ui";
import { Variant, Size } from "./Button.types";

export const variants: Record<Variant, ThemeUIStyleObject> = {
  positive: {
    color: "positiveText",
    background: "positive",
  },
  negative: {
    color: "negativeText",
    background: "negative",
  },
  neutral: {
    color: "neutralText",
    background: "neutral",
  },
  secondary: {
    color: "secondaryText",
    background: "secondary",
  },
  tertiary: {
    color: "tertiaryText",
    background: "tertiary",
  },
};

export const sizes: Record<Size, ThemeUIStyleObject> = {
  sm: {
    height: 40,
    fontSize: "md",
  },
  md: {
    height: 60,
    fontSize: "xl",
  },
};

export const disabledStyle: ThemeUIStyleObject = {
  filter: "brightness(0.6)",
  opacity: 0.8,
  cursor: "unset",
};

export const commonStyle = (disabled: boolean) => ({
  borderRadius: "md",
  boxShadow: "md",
  width: "100%",
  height: 40,
  fontSize: "md",
  fontFamily: "inherit",
  cursor: "pointer",
  transition: "transform 0.1s ease-out",

  ...(!disabled && {
    "&:hover": {
      transform: "scale(1.05)",
      filter: "brightness(1.15)",
    },
    "&:active": {
      transform: "scale(0.98)",
      filter: "brightness(0.95)",
    },
  }),
});
