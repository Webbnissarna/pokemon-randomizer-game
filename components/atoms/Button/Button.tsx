import { Button as StyledButton } from "theme-ui";
import { variants, sizes, disabledStyle, commonStyle } from "./Button.styles";
import { ButtonProps } from "./Button.types";

export default function Button({
  text,
  variant,
  size,
  onClick,
  disabled = false,
}: ButtonProps): JSX.Element {
  return (
    <StyledButton
      sx={{
        ...commonStyle(disabled),
        ...variants[variant],
        ...sizes[size],
        ...(disabled && disabledStyle),
      }}
      disabled={disabled}
      onClick={onClick}
    >
      {text}
    </StyledButton>
  );
}
