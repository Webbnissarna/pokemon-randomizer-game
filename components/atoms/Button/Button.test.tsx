import { fireEvent, screen } from "@testing-library/react";
import { render } from "../../../.jest/render";
import Button from ".";

describe("Button", () => {
  it("renders", () => {
    const { asFragment } = render(
      <Button
        variant="positive"
        size="sm"
        text="Hello World"
        onClick={() => undefined}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("shows text", () => {
    const text = "Hello World";

    render(
      <Button
        variant="positive"
        size="sm"
        text={text}
        onClick={() => undefined}
      />
    );

    const element = screen.getByText(text);

    expect(element).toBeTruthy();
  });

  it("calls onClick", () => {
    const text = "Hello World";
    const callback = jest.fn();

    render(
      <Button variant="positive" size="sm" text={text} onClick={callback} />
    );

    fireEvent.click(screen.getByText(text));

    expect(callback).toHaveBeenCalledTimes(1);
  });

  it("does not call onClick when disabled", () => {
    const text = "Hello World";
    const callback = jest.fn();

    render(
      <Button
        variant="positive"
        size="sm"
        text={text}
        onClick={callback}
        disabled={true}
      />
    );

    fireEvent.click(screen.getByText(text));

    expect(callback).not.toHaveBeenCalled();
  });
});
