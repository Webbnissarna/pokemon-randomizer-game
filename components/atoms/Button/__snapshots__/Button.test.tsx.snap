// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders 1`] = `
<DocumentFragment>
  <script>
    (function() { try {
  var mode = localStorage.getItem('theme-ui-color-mode');
  if (!mode) return
  document.documentElement.classList.add('theme-ui-' + mode);
} catch (e) {} })();
  </script>
  <button
    class="css-4qfqhp-Box"
  >
    Hello World
  </button>
</DocumentFragment>
`;
